src/snapred/backend/data/GroceryService.py

```diff
@@ -190,52 +237,110 @@ def getCloneOfWorkspace(self, name: WorkspaceName, copy: WorkspaceName):
+            #   could also be triggered here.
+
+            wsName = self.uniqueHiddenName()
+
+            # Load the bare instrument:
+            instrumentFilename = (
+                Config["instrument.lite.definition.file"]
+                if useLiteMode
+                else Config["instrument.native.definition.file"]
+            )
+            loadEmptyInstrument = AlgorithmManager.create("LoadEmptyInstrument")
+            loadEmptyInstrument.setProperty("Filename", instrumentFilename)
+            loadEmptyInstrument.setProperty("OutputWorkspace", wsName)
+            loadEmptyInstrument.execute()
+
+            # Initialize the instrument parameters
+            self._updateInstrumentParameters(runId, wsName)
+        self._loadedInstruments[key] = wsName
+        return wsName
+
+    def _updateInstrumentParameters(self, runNumber: str, wsName: str):
         """
-        If the workspace has been loaded, but is not represented in the cache
-        then update the cache to represent this condition
+        Update mutable instrument parameters
         """
-        workspace = self._createRawNeutronWorkspaceName(runId, useLiteMode)
-        key = self._key(runId, useLiteMode)
-        # if the raw data has been loaded but not represent in cache
-        if self.workspaceDoesExist(workspace) and self._loadedRuns.get(key) is None:
-            self._loadedRuns[key] = 0
-        # if the raw data has not been loaded but the cache thinks it has
-        if self._loadedRuns.get(key) is not None and not self.workspaceDoesExist(workspace):
-            del self._loadedRuns[key]
+        # Moved from `PixelGroupingParametersCalculationAlgorithm` for more general application here.
 
-    def _updateGroupingCacheFromADS(self, key, workspace):
+        detectorState: DetectorState = self._getDetectorState(runNumber)
+
+        # Add sample logs with detector "arc" and "lin" parameters to the workspace
+        # NOTE after adding the logs, it is necessary to update the instrument to
+        #  factor in these new parameters, or else calculations will be inconsistent.
+        #  This is done with a call to `ws->populateInstrumentParameters()` from within mantid.
+        #  This call only needs to happen with the last log
+        logsAdded = 0
+        for paramName in ("arc", "lin"):
+            for index in range(2):
+                addSampleLog = AlgorithmManager.create("AddSampleLog")
+                addSampleLog.setProperty("Workspace", wsName)
+                addSampleLog.setProperty("LogName", "det_" + paramName + str(index + 1))
+                addSampleLog.setProperty("LogText", str(getattr(detectorState, paramName)[index]))
+                addSampleLog.setProperty("LogType", "Number Series")
+                addSampleLog.setProperty("UpdateInstrumentParameters", (logsAdded >= 3))
+                addSampleLog.execute()
+                logsAdded += 1
```

Comment:
This would end up being called twice inside PixelGroupingParametersCalculationAlgorithm.  Is there any negative consequence to that, and could that block of code be removed from there in favor of this one?

Reply:
**The "update instrument parameters" code will be removed from `PixelGroupingParametersCalculationAlgorithm`, but that's going to happen in my other story.**  So far, it doesn't appear to throw off the calculation (i.e. all of the "regression" tests are passing).  The key thing with calling it twice is whether or not the location-parameter changes are applied as _relative_ changes or not.  (The fact that the tests continue to pass suggests that they're absolute.)